dwave/embedding/chain_strength.py

```diff
@@ -41,30 +42,36 @@ def uniform_torque_compensation(bqm, embedding=None, prefactor=1.414):
 
     The RMS of the problem's quadratic biases is used for calculation.
 
-    Args: 
+    Args:
         bqm (:obj:`.BinaryQuadraticModel`):
             A binary quadratic model.
 
         embedding (dict/:class:`.EmbeddedStructure`, default=None):
-            Included to satisfy the `chain_strength` callable specifications 
-            for `embed_bqm`. 
+            Included to satisfy the `chain_strength` callable specifications
+            for `embed_bqm`.
 
         prefactor (float, optional, default=1.414):
-            Prefactor used for scaling. For non-pathological problems, the recommended 
+            Prefactor used for scaling. For non-pathological problems, the recommended
             range of prefactors to try is [0.5, 2].
 
     Returns:
         float: The chain strength, or 1 if chain strength is not applicable.
 
     """
-    if bqm.num_interactions > 0:
-        squared_j = (j ** 2 for j in bqm.quadratic.values())
-        rms = math.sqrt(sum(squared_j)/bqm.num_interactions)
+    # NumPy arrays improves performance through vectorization
+    quadratic_array = np.fromiter(bqm.quadratic.values(), dtype=float)
```

Comment:
https://github.com/dwavesystems/dimod/issues/1173 gives some snippets for constructing NumPy matrices from BQMs. Maybe that's worth revisiting.

Also `dwave-system` has SciPy as a dependency. I wonder if we should do this as a sparse matrix instead? And maybe https://github.com/dwavesystems/dimod/issues/1173 should also consider sparse matrices (if SciPy is installed).

Reply:
Are you suggesting using sparse matrix here instead of simply vectorizing the quadratic values? I'm not sure I see how that would improve things unless the quadratic values are very sparse (e.g., doing something like `(sparse_bqm_squared.sum() - sparse_bqm_squared.diagonal().sum()) / num_interactions` for the rms, since only the quadratic values are used).

Reply:
oh, duh, you're doing a 1d matrix of just the values. Ignore me.

Reply:
Though in that case you could use `count=num_interactions`

Reply:
In what way would that help? We'd like to read all quadratic terms anyway, no?

Reply:
It preallocates the numpy vector. Rather than constructing a list first.